devices/src/virtio/queue.rs

```diff
@@ -340,6 +340,44 @@ impl Queue {
         }
     }
 
+    /// Returns the number of yet-to-be-popped descriptor chains in the avail ring.
+    pub fn len(&self, mem: &GuestMemory) -> u16 {
+        (self.avail_idx(mem) - self.next_avail).0
+    }
+
+    /// Checks if the driver has made any descriptor chains available in the avail ring.
+    pub fn is_empty(&self, mem: &GuestMemory) -> bool {
+        self.len(mem) == 0
+    }
+
+    /// Pop the first available descriptor chain from the avail ring.
+    pub fn pop<'a, 'b>(&'a mut self, mem: &'b GuestMemory) -> Option<DescriptorChain<'b>> {
+        if self.len(mem) == 0 {
+            return None;
+        }
+        let index_offset = 4 + 2 * (self.next_avail.0 % self.actual_size());
+
+        // `self.is_valid()` already performed all the bound checks on the descriptor table
+        // and virtq rings, so it's safe to unwrap guest memory reads and to use unchecked
+        // offsets.
+        let desc_index: u16 = mem
+            .read_obj_from_addr(self.avail_ring.unchecked_add(usize::from(index_offset)))
+            .unwrap();
+
+        DescriptorChain::checked_new(mem, self.desc_table, self.actual_size(), desc_index).map(
+            |dc| {
+                self.next_avail += Wrapping(1);
+                dc
+            },
+        )
+    }
+
+    /// Undo the effects of the last `self.pop()` call.
+    /// The caller can use this, if it was unable to consume the last popped descriptor chain.
+    pub fn undo_pop(&mut self) {
```

Comment:
`undo_pop` sounds very unnatural to me. I think I would prefer to do a `peek` and if the operation succeeds do a pop. I think this would lead to some simplifications in the overall flow and fewer error cases.

Reply:
I was (am) pretty much on the fence here, because I do agree that `pop/unpop` is an awkward pattern, and `peek/pop` is better established and hence easier on the eyes. 

The reason I went with `pop/unpop` is that `peek` and `pop` are equally expensive operations, both involving some arithmetic and the invocation of `DescriptorChain::checked_new()`. Doing `peek/pop` would perform all that twice, for each processed virtq buffer. I also considered a `peek/skip` approach, where `skip` would just advance `self.next_avail`, but that didn't seem to address the awkwardness factor either.

All in all, for this particular pickle, I couldn't find a solution with which to be 100% satisfied.